### DISCUSSION TITLE
PIC32MX254F128DT-I/ML

### DISCUSSION THREAD

**Discussion by JacobChrist** [chipKIT32/chipKIT-core#478] [General]
FYI I bought some of the 72MHz PIC32MX254F128DT-I/ML today to see how difficult it will be to get the bootloader to run on a Fubarino Mini PCB.

Jacob

**Comment by majenkotech** [comment]
Theoretically it should be:

* Create a linker script
* Create a set of config bits
* Possibly adjust the CDC/ACM USB code to handle the newer clock settings (not sure what if anything would be needed here)

One part of my work on the new compiler is to automatically generate all the linker scripts we could ever need for us. That includes the bootloader linker script, and the "application" linker script to go along with it.  I haven't got a configuration for the 2x4 chips yet though.

**Comment by JacobChrist** [comment]
@EmbeddedMan @ricklon @majenkotech 
So I got my new boards in and mounted some PIC32MX254F128DT-I/ML parts.  I'm going to start working on a bootloader for them but I wasn't sure what to call it.  The boards are based on the Fubarino Mini 2.0.  My first thought was to call it a Fubarino bootloader but I wasn't sure if that would cause confusion.  If named after the Fubarino I was going to call it FUBARINO_MINI_USB_74MHZ but I was thinking that it might be good to get the chip P/N in the name as well as I might have to switch to the MX274 parts if availability is low on this part.

Any opinions on what this should be called?

![image](https://user-images.githubusercontent.com/198251/150633163-efa1426e-469a-435c-9e38-0a814bfb1625.png)

**Comment by JacobChrist** [comment]
Another issue is that the PIC32MX254F128D is not supported in v1.33 of XC32.  I've installed v4.0 but it appears that the 1+ license doesn't allow for highest level optimizations for this part.  I can use the free version probably for 30 days, but after that if the code doesn't fit in boot I'll need to figure out how to get a license.

**Comment by ricklon** [comment]
Do we have access to the chipKIT name?

If it's too descriptive it will scroll off the important part.

--Rick

On Sat, Jan 22, 2022 at 4:33 AM Jacob Christ ***@***.***>
wrote:

> @EmbeddedMan <https://github.com/EmbeddedMan> @ricklon
> <https://github.com/ricklon> @majenkotech <https://github.com/majenkotech>
> So I got my new boards in and mounted some PIC32MX254F128DT-I/ML parts.
> I'm going to start working on a bootloader for them but I wasn't sure what
> to call it. The boards are based on the Fubarino Mini 2.0. My first thought
> was to call it a Fubarino bootloader but I wasn't sure if that would cause
> confusion. If named after the Fubarino I was going to call it
> FUBARINO_MINI_USB_74MHZ but I was thinking that it might be good to get the
> chip P/N in the name as well as I might have to switch to the MX274 parts
> if availability is low on this part.
>
> Any opinions on what this should be called?
>
> [image: image]
> <https://user-images.githubusercontent.com/198251/150633163-efa1426e-469a-435c-9e38-0a814bfb1625.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/chipKIT32/chipKIT-core/discussions/478#discussioncomment-2021815>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABV3ZBGIIB2I3PGVZ5VNYLUXJ2UZANCNFSM5LLOYPCA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Co-founder
Fair Use Building and Research (FUBAR) Labs
http://fubarlabs.org

**Reply by JacobChrist** [reply]
I was just following the naming of the other existing configurations.  I am primarily asking bemuse just didn't want to cause confusion since there is not a Fubarino that actually would use this chip (at this point).  Also the part is a XLP which could be used in the name.  I haven't dug in too deep but it seems to me that the size of the flash may be important at some point and may affect the configuration but its been so long since I make a linker script or worked on a bootloader that I'm going to be doing a lot of re-learning.

**Comment by EmbeddedMan** [comment]
I could be wrong, but I remember Microchip was going to look into
transferring the chipKIT trademark to 'us', but I never heard any more
about it. My guess is that it might be difficult to get much traction
inside Microchip on the transfer of the trademark. But I could be
remembering this all wrong.

*Brian

**Comment by JacobChrist** [comment]
Woot!  After about 5 hours I got a PIC32MX254F128D bootloader to complier.  I'm 99% sure it will not work as I have yet to go over the config registers and there still may be some linker file issue.

**Comment by JacobChrist** [comment]
The code doesn't fit in the same space as the Fubrino Mini 2.0 and I'm going to chalk this up to not have full optimization.  I had to bump kseg0_program_mem to make it fit.  The chip has 12k boot space, so I'm assuming that this small bump will not cause any issues.

```
MEMORY
{
  kseg1_boot_mem             : ORIGIN = 0xBFC00000, LENGTH = 0x0200
  kseg0_boot_mem             : ORIGIN = 0x9FC00200, LENGTH = 0x0
  kseg0_program_mem    (rx)  : ORIGIN = 0x9FC00200, LENGTH = 0xBF0
  split_program_mem    (rx)  : ORIGIN = 0x9D01F000, LENGTH = 0x1000
  exception_mem              : ORIGIN = 0x9FC01000, LENGTH = 0x0
  debug_exec_mem             : ORIGIN = 0xBFC02000, LENGTH = 0x0

  config3                    : ORIGIN = 0xBFC02FF0, LENGTH = 0x4
  config2                    : ORIGIN = 0xBFC02FF4, LENGTH = 0x4
  config1                    : ORIGIN = 0xBFC02FF8, LENGTH = 0x4
  config0                    : ORIGIN = 0xBFC02FFC, LENGTH = 0x4
  kseg1_data_mem       (w!x) : ORIGIN = 0xA0000000, LENGTH = 0x4000
  sfrs                       : ORIGIN = 0xBF800000, LENGTH = 0x100000
  configsfrs                 : ORIGIN = 0xBFC02FF0, LENGTH = 0x10
/* Starting values from Microship provided .ld file.
  kseg1_boot_mem              : ORIGIN = 0xBFC00000, LENGTH = 0x490
  kseg0_boot_mem              : ORIGIN = 0x9FC00490, LENGTH = 0x970
  kseg0_program_mem     (rx)  : ORIGIN = 0x9D000000, LENGTH = 0x20000
  split_program_mem    (rx)  : ORIGIN = 0x9D01F000, LENGTH = 0x1000
  exception_mem               : ORIGIN = 0x9FC01000, LENGTH = 0x1000
  debug_exec_mem              : ORIGIN = 0xBFC02000, LENGTH = 0xFF0

  config3                     : ORIGIN = 0xBFC02FF0, LENGTH = 0x4
  config2                     : ORIGIN = 0xBFC02FF4, LENGTH = 0x4
  config1                     : ORIGIN = 0xBFC02FF8, LENGTH = 0x4
  config0                     : ORIGIN = 0xBFC02FFC, LENGTH = 0x4
  kseg1_data_mem       (w!x)  : ORIGIN = 0xA0000000, LENGTH = 0x8000
  sfrs                        : ORIGIN = 0xBF800000, LENGTH = 0x100000
  configsfrs                  : ORIGIN = 0xBFC02FF0, LENGTH = 0x10
*/
}

```

**Comment by JacobChrist** [comment]
Opps, these chips require 12MHz or 24MHz crystals for USB to work, but it's a whooping 480Mbps (nice).

**Reply by JacobChrist** [reply]
There is some contradictory information between DS60001250B Section 42 (Oscillators with Enhanced PLL) and  DS60001404E PIC32MX1xx/2xx XLP Family datasheets. Section 42 says a 12MHz or 24MHz Posc is required for HS USB operation (page 42-25), but the XLP Family datasheet indicates that the USB PLL is capable of creating the required 48MHz from an 8MHz Posc by adjusting the UPLLCON SFR (page 83).

Since the UPLLCON SFR is not part of the configuration bits it needs to be done in code, I'm thinking in usb_initialize().  To properly and generically configure these bits two things are required.

- To know the Posc frequency
- To know I that the chip has this capability 

For now this is what I'm thinking.

Add a define for the frequency of Posc like this:
```
#define F_POSC                       8000000UL
```

And use the _CPU_NAME_ to determine if these can be set:
```
#define _CPU_NAME_                  "32MX254F128D"
```

**Reply by JacobChrist** [reply]
Actually right now I'm going to use #if defined( __32MX254F128D_H ) because I'm not sure if v1.33 is C11 and can compare preprocessor strings.

**Reply by JacobChrist** [reply]
Yeah, I got the bootloader to instantiate as a serial port!

![image](https://user-images.githubusercontent.com/198251/150696524-c22c7bdf-8497-49c1-96f5-32cde19796b6.png)

I will not program yet...
```
Sketch uses 28888 bytes (11%) of program storage space. Maximum is 245760 bytes.
Global variables use 7728 bytes (12%) of dynamic memory, leaving 55760 bytes for local variables. Maximum is 63488 bytes.
\\.\COM8: Cannot set state
    Copyright: (C) 2011-2015 Serge Vakulenko

                   2016-2017 Majenko Technologies
No target found.
Programmer for Microchip PIC32 microcontrollers, Version 2.1.46
An error occurred while uploading the sketch
```

**Reply by JacobChrist** [reply]
Yeah, this is strange, the device enumerates and shows up in the device manager as a COM port but it will not open.  The Arduino IDE give this:
```
Error while setting serial port parameters: 115,200 N 8 1
```

**Reply by majenkotech** [reply]
It sounds to me like it's crashing. It'd be interesting to see what Linux's `dmesg` says.

**Reply by JacobChrist** [reply]
```
[ 5881.702685] usb 1-1.3.1.4: new full-speed USB device number 14 using xhci_hcd
[ 5881.828691] usb 1-1.3.1.4: New USB device found, idVendor=0403, idProduct=a662, bcdDevice= 1.80
[ 5881.828697] usb 1-1.3.1.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 5881.828700] usb 1-1.3.1.4: Product: Stk500v2
[ 5881.828703] usb 1-1.3.1.4: Manufacturer: www.cpustick.com
[ 5881.892922] cdc_acm 1-1.3.1.4:1.0: ttyACM0: USB ACM device
[ 5881.893426] usbcore: registered new interface driver cdc_acm
[ 5881.893428] cdc_acm: USB Abstract Control Model driver for USB modems and ISDN adapters
```

It gets a little further in programming:

```
Sketch uses 36116 bytes (14%) of program storage space. Maximum is 245760 bytes.
Global variables use 8444 bytes (13%) of dynamic memory, leaving 55044 bytes for local variables. Maximum is 63488 bytes.
    Copyright: (C) 2011-2015 Serge Vakulenko
                   2016-2017 Majenko Technologies
Programmer for Microchip PIC32 microcontrollers, Version 2.1.46
      Adapter: STK500v2 Bootloader
 Program area: 1d000000-1d1fffff
    Processor: Bootloader
 Flash memory: 2048 kbytes
  Boot memory: 80 kbytes
         Data: 37420 bytes
        Erase: done
Program flash: .......................................Program flash failed.
An error occurred while uploading the sketch
```

No additional dmesg entries.

**Reply by JacobChrist** [reply]
The bootloader LED shut off, but when I cycled power, it went back to the bootloader.

**Comment by JacobChrist** [comment]
das blinkenlichten

**Comment by JacobChrist** [comment]
I think I have System Clock at 72MHz this morning, as the LED blink rate seems to match the Fubrino Mini 2.0 and is now longer blinken every minute.

**Comment by JacobChrist** [comment]
I think I was mistaken.  This chip only has USB 12Mbps (Full Speed).  Is so easy to confuse Full Speed with High Speed (480 Mbps).  I'm actually kind of revealed, I was worried about my USB traces not being controlled impedance and having to verify eyes since I don't have a DCA at my finger tips and the cheapest one on ebay is about $4500.

**Reply by majenkotech** [reply]
I've never really worried about controlled impedance for HS USB. yeah, I try and get my traces pretty much the same length, but I'm not too anal about it, and I've never had any problems with USB.

**Reply by JacobChrist** [reply]
That's good to hear and I was hoping that would be the case but its not even a concern now.  The chip makers need there eyes to be super on spot to deal with shit cables.  I worked in telecom for 8 years testing transceivers and know that if you ignore issues with your eyes they come back to haunt you.  None the less, 12MHz is practically DC and I'm on a 4 layer board with ground planes.

**Reply by JacobChrist** [reply]
Actually, the RF guys I used to work with would say anything less than 1GHz is DC.

**Comment by JacobChrist** [comment]
By the way, I've pushed my code changes on to a branch on my fork:

https://github.com/JacobChrist/PIC32-avrdude-bootloader/tree/PIC32MX254F128D

But there is no configuration in this project for this chip (yet).  I think I jacked up the project and didn't want to push it.  I want to revert and rebuild the configuration before pushing.  All I did though was duplicate the FUBARINO_MINI_50MHz and make a FUBARINO_MINI_72MHz.

**Comment by JacobChrist** [comment]
I opened a support ticket with Microchip on the issue.  At the suggestion of the support engineer I ran a program called usb tree view. So I did so for both the Fubrino Mini 2.0 and the board I'm working on with the PIC32MX254F128D.  There is no difference between the reported data to my PC other than a byte that increments every time a device is enumerated.  (See attached files)

[PIC32MX254F128D-usb-tree-view.txt](https://github.com/chipKIT32/chipKIT-core/files/7929776/PIC32MX254F128D-usb-tree-view.txt)
[fubrino-mini-usb-tree.txt](https://github.com/chipKIT32/chipKIT-core/files/7929777/fubrino-mini-usb-tree.txt)

Also while addressing a question from support I realized that I didn't pull-up pin 41 (USBID) to +3.3V.  I patch a board with this corrected but it didn't help either.

**Reply by majenkotech** [reply]
The USBID is configurable (and usually disabled) in the config bits IIRC. Only used for OTG operation.

**Comment by JacobChrist** [comment]
I just did another pass on the linker script, the only differences now are the config register addresses and an optional include for older (than v2.100) compliers.  Still no beuno.
